File: skillmeat/web/lib/api/context-sync.ts
/**
 * Context sync API service functions
 *
 * API client for bi-directional synchronization of context entities between
 * user collections and deployed projects.
 */

const API_BASE = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8080';
const API_VERSION = process.env.NEXT_PUBLIC_API_VERSION || 'v1';

/**
 * Build API URL with versioned path
 */
function buildUrl(path: string): string {
  return `${API_BASE}/api/${API_VERSION}${path}`;
}

/**
 * Resolution type for sync conflict resolution
 */
export type SyncResolution = 'keep_local' | 'keep_remote' | 'merge';

/**
 * Sync result for a single entity
 */
export interface SyncResult {
  entity_id: string;
  entity_name: string;
  action: 'pulled' | 'pushed' | 'resolved' | 'conflict' | 'skipped';
  message: string;
}

/**
 * Sync conflict between collection and project
 */
export interface SyncConflict {
  entity_id: string;
  entity_name: string;
  entity_type: string;
  collection_hash: string;
  deployed_hash: string;
  collection_content: string;
  deployed_content: string;
  collection_path: string;
  deployed_path: string;
}

/**
 * Sync status response
 */
export interface SyncStatus {
  modified_in_project: string[];
  modified_in_collection: string[];
  conflicts: SyncConflict[];
}

/**
 * Pull changes from project to collection
 *
 * Reads deployed files and updates collection entities with new content.
 */
export async function pullChanges(
  projectPath: string,
  entityIds?: string[]
): Promise<SyncResult[]> {
  const response = await fetch(buildUrl('/context-sync/pull'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      project_path: projectPath,
      entity_ids: entityIds,
    }),
  });

  if (!response.ok) {
    const errorBody = await response.json().catch(() => ({}));
    throw new Error(errorBody.detail || `Failed to pull changes: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Push collection changes to project
 *
 * Writes collection entity content to deployed files.
 */
export async function pushChanges(
  projectPath: string,
  entityIds?: string[],
  overwrite: boolean = false
): Promise<SyncResult[]> {
  const response = await fetch(buildUrl('/context-sync/push'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      project_path: projectPath,
      entity_ids: entityIds,
      overwrite,
    }),
  });

  if (!response.ok) {
    const errorBody = await response.json().catch(() => ({}));
    throw new Error(errorBody.detail || `Failed to push changes: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Get sync status for a project
 *
 * Detects modified entities and conflicts between collection and project.
 */
export async function getSyncStatus(projectPath: string): Promise<SyncStatus> {
  const params = new URLSearchParams({ project_path: projectPath });
  const response = await fetch(buildUrl(`/context-sync/status?${params.toString()}`));

  if (!response.ok) {
    const errorBody = await response.json().catch(() => ({}));
    throw new Error(errorBody.detail || `Failed to get sync status: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Resolve sync conflict
 *
 * Applies user-selected resolution strategy:
 * - keep_local: Update collection from project (project wins)
 * - keep_remote: Update project from collection (collection wins)
 * - merge: Use provided merged_content for both
 */
export async function resolveConflict(
  projectPath: string,
  entityId: string,
  resolution: SyncResolution,
  mergedContent?: string
): Promise<SyncResult> {
  const response = await fetch(buildUrl('/context-sync/resolve'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      project_path: projectPath,
      entity_id: entityId,
      resolution,
      merged_content: mergedContent,
    }),
  });

  if (!response.ok) {
    const errorBody = await response.json().catch(() => ({}));
    throw new Error(errorBody.detail || `Failed to resolve conflict: ${response.statusText}`);
  }

  return response.json();
}

File: skillmeat/api/schemas/context_sync.py
"""Context sync API request/response schemas.

Pydantic models for context entity synchronization endpoints.
"""

from typing import List, Literal, Optional

from pydantic import BaseModel, Field


class SyncPullRequest(BaseModel):
    """Request schema for pulling changes from project to collection.

    Attributes:
        project_path: Absolute path to project directory
        entity_ids: Optional list of entity IDs to pull (pulls all if None)
    """

    project_path: str = Field(
        ...,
        description="Absolute path to project directory",
        examples=["/path/to/project"],
    )
    entity_ids: Optional[List[str]] = Field(
        None,
        description="Optional list of entity IDs to pull (pulls all if None)",
        examples=[["spec_file:api-patterns", "rule_file:debugging"]],
    )


class SyncPushRequest(BaseModel):
    """Request schema for pushing collection changes to project.

    Attributes:
        project_path: Absolute path to project directory
        entity_ids: Optional list of entity IDs to push (pushes all if None)
        overwrite: If True, push even if file modified locally (force)
    """

    project_path: str = Field(
        ...,
        description="Absolute path to project directory",
        examples=["/path/to/project"],
    )
    entity_ids: Optional[List[str]] = Field(
        None,
        description="Optional list of entity IDs to push (pushes all if None)",
        examples=[["spec_file:api-patterns"]],
    )
    overwrite: bool = Field(
        False,
        description="If True, push even if file modified locally (force)",
    )


class SyncConflictResponse(BaseModel):
    """Response schema for sync conflict information.

    Attributes:
        entity_id: Entity identifier
        entity_name: Entity name
        entity_type: Entity type (spec_file, rule_file, etc.)
        collection_hash: Hash in collection
        deployed_hash: Hash of deployed file
        collection_content: Current content in collection
        deployed_content: Current content in deployed file
        collection_path: Path to entity in collection
        deployed_path: Path to deployed file in project
        change_origin: Origin of the change (optional)
        baseline_hash: Hash at deployment time (baseline for three-way merge)
    """

    entity_id: str = Field(..., description="Entity identifier")
    entity_name: str = Field(..., description="Entity name")
    entity_type: str = Field(..., description="Entity type")
    collection_hash: str = Field(..., description="Content hash in collection")
    deployed_hash: str = Field(..., description="Content hash of deployed file")
    collection_content: str = Field(..., description="Current content in collection")
    deployed_content: str = Field(..., description="Current content in deployed file")
    collection_path: str = Field(..., description="Path to entity in collection")
    deployed_path: str = Field(..., description="Path to deployed file in project")
    change_origin: Optional[str] = Field(
        default=None,
        description="Origin of the change (deployment/sync/local_modification)",
    )
    baseline_hash: Optional[str] = Field(
        default=None,
        description="Hash at deployment time (baseline for three-way merge)",
    )


class SyncResultResponse(BaseModel):
    """Response schema for sync operation result.

    Attributes:
        entity_id: Entity identifier
        entity_name: Entity name
        action: Action performed (pulled, pushed, skipped, conflict, resolved)
        message: Human-readable status message
    """

    entity_id: str = Field(..., description="Entity identifier")
    entity_name: str = Field(..., description="Entity name")
    action: str = Field(
        ...,
        description="Action performed",
        examples=["pulled", "pushed", "skipped", "conflict", "resolved"],
    )
    message: str = Field(..., description="Human-readable status message")


class SyncStatusResponse(BaseModel):
    """Response schema for sync status information.

    Attributes:
        modified_in_project: Entity IDs modified in project
        modified_in_collection: Entity IDs modified in collection
        conflicts: List of sync conflicts
    """

    modified_in_project: List[str] = Field(
        ...,
        description="Entity IDs modified in project",
        examples=[["spec_file:api-patterns"]],
    )
    modified_in_collection: List[str] = Field(
        ...,
        description="Entity IDs modified in collection",
        examples=[["rule_file:debugging"]],
    )
    conflicts: List[SyncConflictResponse] = Field(
        ..., description="List of sync conflicts"
    )


class SyncResolveRequest(BaseModel):
    """Request schema for resolving sync conflicts.

    Attributes:
        project_path: Absolute path to project directory
        entity_id: Entity identifier to resolve
        resolution: Resolution strategy (keep_local, keep_remote, merge)
        merged_content: Required if resolution is "merge"
    """

    project_path: str = Field(
        ...,
        description="Absolute path to project directory",
        examples=["/path/to/project"],
    )
    entity_id: str = Field(
        ...,
        description="Entity identifier to resolve",
        examples=["spec_file:api-patterns"],
    )
    resolution: Literal["keep_local", "keep_remote", "merge"] = Field(
        ...,
        description="Resolution strategy",
    )
    merged_content: Optional[str] = Field(
        None,
        description="Required if resolution is 'merge'",
    )
